BUG: Preserve custom color tables on scene view restore

naucoin opened the pull request:
Moving the color table information into files on disk created
a bug (analogous to the same situation with fiducials[1]) that when
a scene contains scene views that contain custom color tables
the scene view restore wipes out the colors in the color node
in the main scene. When a MRML file is read, the files on disk
are not reread to fill in scene view nodes, so the scene view
color table node is empty. This fix adds a check for if the
scene is currently restoring a scene view, and if so, don't
proceed with the copy as it will result in an color table
with the correct number of entries (as the number of colors is
saved in the MRML file) but all black and unamed.

[1] fiducials were fixed in this commit:
http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=24376
https://github.com/Slicer/Slicer/commit/1440084bbdd9d11c1e9e1f195087f521a91af186

Issue #3992

jcfr commented:
Nice.

In your commit message, you mention `[...] analagous to the same situation with fiducials [...]`, could you:
* `analagous` -> `analogous`
* reference the commit message that fixed the problem for fiducial.

naucoin commented:
Comment updated.

naucoin commented:
I updated the branch with a test, squashed two commits and rebased.

@jcfr  your question about skipping the copy: built in color nodes are not saved to scene views, they're singletons, so they won't get copied during a scene view restore. They need to be copied when making a user editiable version of a built in color node.

jcfr commented:
```
$ ctest -R Color 
Test project /home/jcfr/Projects/Slicer-Debug/Slicer-build
      Start  11: vtkMRMLColorNodeTest1
 1/28 Test  #11: vtkMRMLColorNodeTest1 ...............................   Passed    0.82 sec
      Start  12: vtkMRMLColorTableNodeTest1
 2/28 Test  #12: vtkMRMLColorTableNodeTest1 ..........................   Passed    0.88 sec
      Start  13: vtkMRMLColorTableStorageNodeTest1
 3/28 Test  #13: vtkMRMLColorTableStorageNodeTest1 ...................   Passed    0.79 sec
      Start  15: vtkMRMLdGEMRICProceduralColorNodeTest1
 4/28 Test  #15: vtkMRMLdGEMRICProceduralColorNodeTest1 ..............   Passed    0.80 sec
      Start  30: vtkMRMLFreeSurferProceduralColorNodeTest1
 5/28 Test  #30: vtkMRMLFreeSurferProceduralColorNodeTest1 ...........   Passed    0.86 sec
      Start  53: vtkMRMLPETProceduralColorNodeTest1
 6/28 Test  #53: vtkMRMLPETProceduralColorNodeTest1 ..................   Passed    0.83 sec
      Start  54: vtkMRMLProceduralColorNodeTest1
 7/28 Test  #54: vtkMRMLProceduralColorNodeTest1 .....................   Passed    0.82 sec
      Start  55: vtkMRMLProceduralColorStorageNodeTest1
 8/28 Test  #55: vtkMRMLProceduralColorStorageNodeTest1 ..............   Passed    0.79 sec
      Start 118: vtkMRMLColorLogicTest1
 9/28 Test #118: vtkMRMLColorLogicTest1 ..............................   Passed    1.74 sec
      Start 141: qMRMLColorListViewTest1
10/28 Test #141: qMRMLColorListViewTest1 .............................   Passed    1.57 sec
      Start 142: qMRMLColorModelTest
11/28 Test #142: qMRMLColorModelTest .................................   Passed    1.11 sec
      Start 143: qMRMLColorModelTest1
12/28 Test #143: qMRMLColorModelTest1 ................................   Passed    1.37 sec
      Start 144: qMRMLColorPickerWidgetTest1
13/28 Test #144: qMRMLColorPickerWidgetTest1 .........................   Passed    1.86 sec
      Start 145: qMRMLColorPickerWidgetTest2
14/28 Test #145: qMRMLColorPickerWidgetTest2 .........................   Passed    1.44 sec
      Start 146: qMRMLColorPickerWidgetTest3
15/28 Test #146: qMRMLColorPickerWidgetTest3 .........................   Passed    1.53 sec
      Start 147: qMRMLColorTableComboBoxTest1
16/28 Test #147: qMRMLColorTableComboBoxTest1 ........................   Passed    1.15 sec
      Start 148: qMRMLColorTableViewTest1
17/28 Test #148: qMRMLColorTableViewTest1 ............................   Passed    1.43 sec
      Start 179: qMRMLSceneColorTableModelTest1
18/28 Test #179: qMRMLSceneColorTableModelTest1 ......................   Passed    1.41 sec
      Start 206: qMRMLColorListViewEventTranslatorPlayerTest1
19/28 Test #206: qMRMLColorListViewEventTranslatorPlayerTest1 ........   Passed    4.43 sec
      Start 207: qMRMLColorPickerWidgetEventTranslatorPlayerTest1
20/28 Test #207: qMRMLColorPickerWidgetEventTranslatorPlayerTest1 ....   Passed    4.17 sec
      Start 208: qMRMLColorTableComboBoxEventTranslatorPlayerTest1
21/28 Test #208: qMRMLColorTableComboBoxEventTranslatorPlayerTest1 ...   Passed    1.40 sec
      Start 209: qMRMLColorTableViewEventTranslatorPlayerTest1
22/28 Test #209: qMRMLColorTableViewEventTranslatorPlayerTest1 .......   Passed    4.61 sec
      Start 275: qSlicerColorsModuleGenericTest
23/28 Test #275: qSlicerColorsModuleGenericTest ......................   Passed    3.88 sec
      Start 276: qSlicerColorsModuleWidgetGenericTest
24/28 Test #276: qSlicerColorsModuleWidgetGenericTest ................   Passed    4.19 sec
      Start 277: vtkSlicerColorLogicTest1
25/28 Test #277: vtkSlicerColorLogicTest1 ............................   Passed    1.43 sec
      Start 278: qSlicerColorsModuleWidgetTest1
26/28 Test #278: qSlicerColorsModuleWidgetTest1 ......................   Passed    5.78 sec
      Start 279: py_ColorsScalarBarSelfTest
27/28 Test #279: py_ColorsScalarBarSelfTest ..........................   Passed   90.68 sec
      Start 280: py_CustomColorTableSceneViewRestoreTestBug3992
28/28 Test #280: py_CustomColorTableSceneViewRestoreTestBug3992 ......   Passed    7.98 sec

100% tests passed, 0 tests failed out of 28

Label Time Summary:
MRMLCore                      =   6.58 sec
MRMLLogic                     =   1.74 sec
qMRMLWidgets                  =  27.46 sec
qSlicerColorsModule           =  13.85 sec
vtkSlicerColorsModuleLogic    =   1.43 sec

Total Test time (real) = 149.81 sec
```

jcfr commented:
Integrated in [r24665](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=24665)

jcfr commented:
:+1:
